// webpack-config/env-config.js
// 建议总是将前端静态资源按一个整体 serve 出去
// 相应的，流控可以简单配置成 webapp-domain/static/* -> frontend-domain/*
module.exports = {
    dev: {
        IS_DEV: false,
        APP_PROFILE: 'dev',
        API_CONTEXT: '',
        WEB_CONTEXT: 'static',
    },
    sit: {
        IS_DEV: false,
        APP_PROFILE: 'dev',
        API_CONTEXT: '',
        WEB_CONTEXT: 'static',
    },
    uat: {
        IS_DEV: false,
        APP_PROFILE: 'dev',
        API_CONTEXT: '',
        WEB_CONTEXT: 'static',
    },
    pot: {
        IS_DEV: false,
        APP_PROFILE: 'prod',
        API_CONTEXT: '',
        WEB_CONTEXT: 'static',
    },
    prd: {
        IS_DEV: false,
        APP_PROFILE: 'prod',
        API_CONTEXT: '',
        WEB_CONTEXT: 'static',
    },
    backend: {
        IS_DEV: false,
        APP_PROFILE: '@application.profile@',
        API_CONTEXT: '',
        WEB_CONTEXT: 'static',
    },
    standalone: {
        IS_DEV: false,
        APP_PROFILE: '',
        API_CONTEXT: '',
        WEB_CONTEXT: '',
    }
}[process.env.WEBPACK_ENV] || {
    IS_DEV: true,
    APP_PROFILE: '',
    API_CONTEXT: '',
    WEB_CONTEXT: '',
}
