// webpack-config/frontend-entries.js
/**
 * @file frontend-entries
 * @author lzheng
 */
var utils = require('common/utils')
var _ = require('lodash')

module.exports = function (config) {

    var templateWeb = {
        dev: utils.p(config.path.webpackTemplates + '/dev-page.html'),
        prod: utils.p(config.path.webpackTemplates + '/prod-page.html'),
    };

    var templateMobileApproval = {
        dev: utils.p(config.path.webpackTemplates + '/mobile-approval-dev-page.html'),
        prod: utils.p(config.path.webpackTemplates + '/mobile-approval-prod-page.html'),
    };

    var templateMobileStandalone = {
        dev: utils.p(config.path.webpackTemplates + '/mobile-standalone-dev-page.html'),
        prod: utils.p(config.path.webpackTemplates + '/mobile-standalone-prod-page.html'),
    };

    var templateMobileGeneral = {
        dev: utils.p(config.path.webpackTemplates + '/mobile-general-dev-page.html'),
        prod: utils.p(config.path.webpackTemplates + '/mobile-general-prod-page.html')
    };

    var entries = {}

    // 业务模块入口
    // 注意调整 alias
    _.merge(entries, {
        main: {
            entry: 'frontend/module/MainModule/index',
            template: templateWeb
        },

        example: {
            entry: 'frontend/module/ExampleModule/index',
            template: templateWeb
        },


        mobile: {
            entry: 'frontend/module/TestMobile/index',
            type: 'mobile-general',
            // 是否使用路由, 不使用则vendor中不会引入react-route
            useRouter: true,
            // 是否使用svg symbol, 不使用svg与其他图片一样处理
            useSvgSymbol: true,
            // 可选, 使用svg-sprite-loader加载的icon路径
            svgPath: ['componentMobile/Icon/icons'],
            template: templateMobileGeneral
        },

        'ui-demo-shared': {
            entry: 'frontend/module/UIDemoModule/uiDemoModules/index-shared',
            template: templateWeb
        },

        'ui-demo-frontend': {
            entry: 'frontend/module/UIDemoModule/uiDemoModules/index-frontend',
            template: templateWeb
        },
    });

    return entries
}
