// webpack-config/frontend-entries.js
/**
 * @file frontend-entries
 * @author lzheng
 */
var utils = require('common/utils')
var _ = require('lodash')

module.exports = function (config) {

    var templateWeb = {
        dev: utils.p(config.path.webpackTemplates + '/dev-page.html'),
        prod: utils.p(config.path.webpackTemplates + '/prod-page.html'),
    };

    var entries = {}

    // 业务模块入口
    // 注意调整 alias
    _.merge(entries, {
        main: {
            entry: 'frontend/module/MainModule/index',
            template: templateWeb
        },

        example: {
            entry: 'frontend/module/ExampleModule/index',
            template: templateWeb
        },

        'ui-demo-shared': {
            entry: 'frontend/module/UIDemoModule/uiDemoModules/index-shared',
            template: templateWeb
        },

        'ui-demo-frontend': {
            entry: 'frontend/module/UIDemoModule/uiDemoModules/index-frontend',
            template: templateWeb
        },

        index: {
            entry: 'frontend/module/IndexModule/index',
            template: templateWeb
        },
    });

    return entries
}
