// webpack-config/plugins/copy.js
/**
 * @file: copy
 * @author: Cuttle Cong
 * @date: 2018/1/4
 * @description:
 */
var CopyWebpackPlugin = require('copy-webpack-plugin')
var minify = require('minify')
var config = require('common/config')
var nps = require('path')

var array = [
    {
        from: {
            glob: nps.join(config.path.extraFrontendAsset, '**/*'), dot: false
        },
        // relative path
        context: config.path.extraFrontendAsset,
        to: '../extra',
        toType: 'dir',
        // cache: true,
        transform: function (body, from) {
            var ext = from.replace(/.+\.(.+?)$/, '$1')
            var isI18n = /\/i18n\//.test(from) && ext === 'js'

            return new Promise(function (resolve, reject) {
                var method = minify[ext]
                if (!method) {
                    resolve(body)
                    return
                }
                method(body.toString(), function (error, data) {
                    if (error) {
                        reject(error)
                        console.error(error)
                        process.exit(1)
                    }
                    if (isI18n) {
                        data = data + '\nlang || _lang;'
                    }
                    if (ext === 'js') {
                        data = '/* eslint-disable */\n' + data
                    }
                    resolve(data)
                })
            })
        }
    }
]

module.exports = new CopyWebpackPlugin(array, { debug: null })
