// webpack-config/dll-config/mobile-general-dll-webpack-config-prod.js
/**
 * @file frontend-entries
 * @author lzheng
 */
var config = require('common/config')
var utils = require('common/utils')
var webpack = require('webpack')
var getLoader = require('../get-entry-webpack-config/loader-base/index');
var BundleAnalyzerPlugin = require('webpack-bundle-analyzer')
    .BundleAnalyzerPlugin

var webpackConfig = {
    context: utils.p(config.path.src),
    mode: 'production',
    entry: {
        // vendor 由于要提供给 dll plugin 做 global name, 所以必须是 js 命名友好
        'mobile_general_vendor': [
            'react', 'react-dom', 'react-router',
            'superagent',
            'classnames',

            'mobx',
            'mobx-react',

            'bluebird',
            'core-js'
        ]
    },

    optimization: {
        minimize: true
    },

    module: {
        rules: getLoader({
            isDev: false,
            useSvgSymbol: true
        })
    },

    output: {
        path: config.path.dllProd,
        filename: '[name].bundle.js',

        // The name of the global variable which the library's
        // require() function will be assigned to
        library: '[name]_lib'
    },

    plugins: [
        new webpack.DefinePlugin({
            '__PRODUCTION__': true,
            '__DEVELOPMENT__': false,
            '__DEVTOOLS__': false
        }),
        new webpack.DllPlugin({
            context: utils.p(config.path.src),

            // The path to the manifest file which maps between
            // modules included in a bundle and the internal IDs
            // within that bundle
            path: utils.p(config.path.dllProd + '/[name]-manifest.json'),

            // The name of the global variable which the library's
            // require function has been assigned to. This must match the
            // output.library option above
            name: '[name]_lib'
        }),
        // new BundleAnalyzerPlugin({
        //     analyzerMode: 'static',
        //     reportFilename: 'report_mobile_vendor.html'
        // }),
    ]
}

module.exports = webpackConfig
