// webpack-config/dll-config/pc-normal-dll-webpack-config-dev.js
/**
 * @file dll.webpack.config.dev
 * @author lzheng
 */
var config = require('common/config')
var utils = require('common/utils')
var webpack = require('webpack')

var webpackConfig = {
    devtool: 'cheap-source-map',
    context: utils.p(config.path.src),
    mode: 'development',
    entry: {
        vendor: [
            'react', 'react-dom', 'superagent', 'classnames', 'react-router',

            'mobx',

            'mobx-react',
            'moment', 'bluebird',
            // 'core-js'
        ]
    },

    module: {
        rules: require('../get-entry-webpack-config/loader-base/index')({
            isDev: true
        })
    },

    output: {
        path: config.path.dllDev,
        filename: '[name].bundle.js',

        // The name of the global variable which the library's
        // require() function will be assigned to
        library: '[name]_lib'
    },

    plugins: [
        new webpack.DefinePlugin({
            '__PRODUCTION__': false,
            '__DEVELOPMENT__': true,
            '__DEVTOOLS__': true
        }),
        new webpack.DllPlugin({
            context: utils.p(config.path.src),

            // The path to the manifest file which maps between
            // modules included in a bundle and the internal IDs
            // within that bundle
            path: utils.p(config.path.dllDev + '/[name]-manifest.json'),

            // The name of the global variable which the library's
            // require function has been assigned to. This must match the
            // output.library option above
            name: '[name]_lib'
        }),
        new webpack.ContextReplacementPlugin(
            /moment[\/\\]locale$/,
            /zh-cn/
        ),
    ]
}

module.exports = webpackConfig
